### PR TITLE
Make mpi (device-aware) examples backend agnostic

### PR DESCRIPTION
**Pull request by JackAKirk**
SYCL + MPI should work in this way for any backend that supports MPI aware "device" memory.

We know that these samples work on cuda/hip backends, they are supposed to work on L0, and indeed in theory should also work when the device memory corresponds to cpu memory (e.g. opencl cpu/native cpu backends).

### CODE REVIEW

File: src/MPI_with_SYCL/CMakeLists.txt
Comment (rafbiels) [inline src/MPI_with_SYCL/CMakeLists.txt:1]: This file needs to be adjusted to build the example regardless of `ENABLE_CUDA`.

Could you please also adjust the README section about this example?

File: src/MPI_with_SYCL/send_recv_buff.cpp
Comment (rafbiels) [inline src/MPI_with_SYCL/send_recv_buff.cpp:86]: Does this build for you? The [backend enum](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/backend_types.hpp) values are:
`ext_oneapi_hip`, `ext_oneapi_level_zero`, `ext_oneapi_cuda`

File: src/MPI_with_SYCL/send_recv_buff.cpp
Comment (rafbiels) [inline src/MPI_with_SYCL/send_recv_buff.cpp:78]: Why use macros? Couldn't we just do:
```cpp
if (q.get_backend()==sycl::backend::ext_oneapi_hip)
```
etc? Actually, thinking now that this might need to be still protected with `#if SYCL_EXT_ONEAPI_BACKEND_HIP` because possibly `get_native_mem<sycl::backend::hip>` won't compile without the plugin installed - please check.

Regardless, I believe no new macros need to be introduced.

**Review by rafbiels — CHANGES_REQUESTED** [review]

File: src/MPI_with_SYCL/send_recv_buff.cpp
Comment (JackAKirk) [reply src/MPI_with_SYCL/send_recv_buff.cpp:78]: I think the problem is that `q.get_backend()` etc can't be called inside a kernel, so that I'd have to have two versions of the kernel? Not sure if you know a better way?

Comment (rafbiels) [reply src/MPI_with_SYCL/send_recv_buff.cpp:78]: Would it work to store the value before the kernel and capture it in the lambda?

File: src/MPI_with_SYCL/CMakeLists.txt
Comment (JackAKirk) [reply src/MPI_with_SYCL/CMakeLists.txt:1]: Do you mean to make it build for other backends than CUDA?

File: src/MPI_with_SYCL/send_recv_buff.cpp
Comment (JackAKirk) [reply src/MPI_with_SYCL/send_recv_buff.cpp:78]: It turns out that unfortunately it is not currently possible to write this without using the MACROs in some way regardless. There is an `interop_handle::get_backend()` that would be ideal. However `get_native_mem<sycl::backend::ext_oneapi_cuda>` cannot be compiled when DPC++ is not built with the cuda backend, and `get_native_mem<sycl::backend::ext_oneapi_hip>` cannot be compiled when DPC++ is not build with the hip backend.
In order to achieve true portable interop with buffers `get_native_mem` needs to be invocable without a template parameter: essentially I think the impl would then be changed to use compiler macros instead of user macros to select the correct native backend type. Really this needs a SYCL spec change also to remove the template parameter.

The current situation is that we can either use the MACROs or remove the buffer example, or make it only for one backend.

File: src/MPI_with_SYCL/send_recv_buff.cpp
Comment (JackAKirk) [reply src/MPI_with_SYCL/send_recv_buff.cpp]: I've fixed this now thanks. Apparently they are changing back to `cuda` etc but apparently not yet.

File: src/MPI_with_SYCL/CMakeLists.txt
Comment (JackAKirk) [reply src/MPI_with_SYCL/CMakeLists.txt:1]: Does the CI have a MPI module such that `make` actually compiles the program? When I do it locally without a cmake module it executes `make` without complaining but doesn't do anything.

Comment (JackAKirk) [reply src/MPI_with_SYCL/CMakeLists.txt:1]: I'll test cmake with rocm-aware MPI and see if it can work.

File: src/MPI_with_SYCL/send_recv_buff.cpp
Comment (rafbiels) [reply src/MPI_with_SYCL/send_recv_buff.cpp:78]: Thanks, I was afraid this'd be the case. But it should still be possible to use the DPC++ macros like `SYCL_EXT_ONEAPI_BACKEND_HIP`, right? Then the user doesn't need to provide any extra options for compilation, the preprocessor will include the right sections for them based on the plugins availability. And for the runtime we can still use `queue::get_backend()` to pick the right `get_native_mem` template.

Comment (JackAKirk) [reply src/MPI_with_SYCL/send_recv_buff.cpp:78]: e.g. SYCL_EXT_ONEAPI_BACKEND_HIP is not defined by default. I could replace e.g. USE_HIP with SYCL_EXT_ONEAPI_BACKEND_HIP if you want? It basically just enables a few "extension" interop features that aren't used in this code.

I think maybe you are thinking of using
```
if defined(__SYCL_DEVICE_ONLY__) && defined(__NVPTX__)
```
which are compiler flags which you are right can be used to distinguish between backends in kernel code. However in interop host tasks whilst device memory addresses are exposed from the buffer, the execution is still on host such that these macros are not defined.

I don't think there is anyway around the user defining some macro currently. I can use whatever macros you think best.
I don't really understand why you would still use `queue::get_backend()` in this case. But I don't mind how this is implemented. If you tell me how you want to do it I'll update it.
Thanks

Comment (rafbiels) [reply src/MPI_with_SYCL/send_recv_buff.cpp:78]: Here's what I meant: https://github.com/JackAKirk/SYCL-samples/pull/1

This should compile in any combination of NVIDIA/AMD plugins installed or not installed and then should pick the right template at runtime depending on the device used by the queue. This also allows multi-target compilation where the right option is chosen depending on `ONEAPI_DEVICE_SELECTOR`.

The macros are defined here:  
https://github.com/intel/llvm/blob/sycl/sycl/plugins/cuda/include/features.hpp  
https://github.com/intel/llvm/blob/sycl/sycl/plugins/hip/include/features.hpp  
and this is what decides on the availability of the corresponding get_native_mem template instantiations at compile time. These macros are 1 in open-source but are both 0 in oneAPI release. Installing the NVIDIA or AMD plugin from Codeplay changes the corresponding macro to 1.

Comment (JackAKirk) [reply src/MPI_with_SYCL/send_recv_buff.cpp]: Nice, I see what you mean now. Sorry I didn't realize those macros were now set already.
I've merged your changes (I just made a small change to get the backend directly from interop_handle)

I've updated the readme. I've tested the CMake on Crusher/Perlmutter. It does work, although in both cases you have to manually include/link the correct headers/libs, since CMAKE finds the wrong MPI by default. I tried to find a way to encourage it to find the correct ones. I could get it to find the right lib using CMAKE_PREFIX_PATH, but the only way to get it to find the right include was to set it manually, so I think the best way is to just directly edit the cmake in such a case. I've documented this.
I also mostly replaced "GPU-aware" with "device-aware", and added a discussion to explain this (see the readme). Whilst I tested the buffer test and it cannot yet work with opencl/native_cpu backends because the buffer interop isn't implemented in this case for native_cpu (and in opencl presumably it returns an opencl buffer rather than a pointer), the USM samples should work with CPU devices also. In theory this could be quite useful for people that want portable code across CPU/GPU systems.

Let me know what you think.

Comment (JackAKirk) [reply src/MPI_with_SYCL/send_recv_buff.cpp]: @rafbiels Do you know what clang-format settings are used in the CI? It doesn't give me a diff, and it is clearly different to the settings I have for clang-format for intel/llvm. No worries if not, I'll just do trial and error.

File: src/MPI_with_SYCL/send_recv_buff.cpp
Comment (JackAKirk) [reply src/MPI_with_SYCL/send_recv_buff.cpp:9]: Yeah I just spotted this. There's still a few small things like this I'll correct.

File: src/MPI_with_SYCL/send_recv_buff.cpp
Comment (rafbiels) [reply src/MPI_with_SYCL/send_recv_buff.cpp]: Many thanks @JackAKirk, that looks good to me now. Thanks a lot for this work!

The clang-format in CI uses the config included in the repo:
https://github.com/codeplaysoftware/SYCL-samples/blob/main/.clang-format
and it's specifically running clang-format release 17

File: src/MPI_with_SYCL/CMakeLists.txt
Comment (JackAKirk) [inline src/MPI_with_SYCL/CMakeLists.txt:5]: needs update

File: README.md
Comment (JackAKirk) [inline README.md:76]: needs update

File: src/MPI_with_SYCL/send_recv_buff.cpp
Comment (JackAKirk) [reply src/MPI_with_SYCL/send_recv_buff.cpp]: OK thanks. I'll fix it all up and correct the mentions of cuda specific stuff etc.

Comment (rafbiels) [reply src/MPI_with_SYCL/send_recv_buff.cpp]: I'll be off until 2 April, but consider this approved from my side after those remaining small corrections. Ask @DuncanMcBain to merge once those are done.

File: src/MPI_with_SYCL/send_recv_buff.cpp
Comment (rafbiels) [inline src/MPI_with_SYCL/send_recv_buff.cpp:110]:
```
/__w/SYCL-samples/SYCL-samples/src/MPI_with_SYCL/send_recv_buff.cpp:110:22: error: unused variable 'status' [-Werror,-Wunused-variable]
  110 |           MPI_Status status;
      |                      ^~~~~~
```
The issue was there before, but we weren't building this example in CI (because it required CUDA). Would you mind removing the unused variable?

Comment (JackAKirk) [reply src/MPI_with_SYCL/send_recv_buff.cpp]: Cheers, done

**Review by rafbiels — APPROVED** [review]

**Pull request merged**
